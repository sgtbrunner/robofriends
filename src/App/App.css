@font-face {
  font-family: 'SEGA LOGO FONT';
  font-style: normal;
  font-weight: normal;
  src: local('SEGA LOGO FONT'), url('../assets/SEGA.woff') format('woff');
}

h3,
.t1 {
  color: white;
}

.t1 {
  font-family: 'SEGA LOGO FONT';
  font-weight: 200;
  font-size: 4em;
}

.alert-container {
  height: 100vh;
  display: flex;
  flex-direction: column;
  justify-content: center;
}
