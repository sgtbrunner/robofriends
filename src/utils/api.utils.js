export const USERS_URL = 'https://jsonplaceholder.typicode.com/users';

export default {
  getUsers: () =>
    fetch(USERS_URL).then((response) => {
      if (response.ok) return response.json();
      throw new Error('Something went wrong');
    }),
};
