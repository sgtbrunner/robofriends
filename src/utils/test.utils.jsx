import React from 'react';
import { render } from '@testing-library/react';

export const mountComponent = ({ component, props = {} }) => {
  const Component = component;
  return render(<Component {...props} />);
};

export default null;
