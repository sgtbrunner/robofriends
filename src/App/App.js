import React, { Component } from 'react';
import Cardlist from '../components/Cardlist';
import SearchBox from '../components/searchbox';
import Scroll from '../components/scroll';
import './App.css';

class App extends Component {
  constructor() {
    super();
    this.state = {
      robots: [],
      searchField: '',
    };
  }

  componentDidMount() {
    fetch('https://jsonplaceholder.typicode.com/users')
      .then((response) => response.json())
      .then((users) => this.setState({ robots: users }));
  }

	onSearchChange = (event) => {
	  this.setState({ searchField: event.target.value });
	};

	filterRobots = () => {
	  return this.state.robots.filter((robot) => {
	    return robot.name.toLowerCase().includes(this.state.searchField.toLowerCase());
	  });
	};

	render() {
	  if (!this.state.robots.length) {
	    return <h1 className="tc f1">Loading...</h1>;
	  }
	  return (
  <div className="tc ">
    <h1 className="f2">RoboFriends</h1>
    <SearchBox searchChange={this.onSearchChange} />
    <Scroll>
      <Cardlist robots={this.filterRobots()} />
    </Scroll>
  </div>
	  );
	}
}

export default App;
