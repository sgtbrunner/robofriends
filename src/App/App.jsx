import React, { Component } from 'react';

import Cardlist from '../components/Cardlist/cardlist.component';
import ErrorBoundary from '../components/Error/error-boundary.component';
import Loading from '../components/Loading/loading.component';
import SearchBox from '../components/Searchbox/searchbox.component';
import './App.css';

const USERS_URL = 'https://jsonplaceholder.typicode.com/users';

class App extends Component {
  constructor() {
    super();
    this.state = {
      error: null,
      loading: false,
      robots: [],
      searchField: '',
    };
  }

  componentDidMount() {
    this.setState({ loading: true });
    fetch(USERS_URL)
      .then((response) => response.json())
      .then((users) => this.setState({ robots: users }))
      .catch((error) => this.setState({ error }))
      .finally(() => this.setState({ loading: false }));
  }

  _filterRobots = () =>
    this.state.robots.filter((robot) => {
      return robot.name.toLowerCase().includes(this.state.searchField.toLowerCase());
    });

  _onSearchChange = (event) => {
    this.setState({ searchField: event.target.value });
  };

  render() {
    const { error, loading } = this.state;
    const { _filterRobots, _onSearchChange } = this;

    return (
      <div className="tc">
        {(loading || error) && (
          <div className="message-container">
            {loading && <Loading />}
            {error && <ErrorBoundary />}
          </div>
        )}
        {!error && !loading && (
          <div>
            <h1 className="t1">RoboFriends</h1>
            <SearchBox searchChange={_onSearchChange} />
            <div className="pv4">
              <Cardlist robots={_filterRobots()} />
            </div>
          </div>
        )}
      </div>
    );
  }
}

export default App;
