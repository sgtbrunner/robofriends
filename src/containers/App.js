import React, {Component} from 'react';
import Cardlist from '../components/cardlist';
import SearchBox from '../components/searchbox';
import Scroll from '../components/scroll'
import './App.css';

class App extends Component {
	// STATE DECLARATION
	constructor() {
		super()
		this.state = {
			robots: [],
			searchField: ''
		}
	}

	// componentDidMount() IS A LIFECYCLE HOOK THAT RUNS ONCE THE APP COMPONENT IS MOUNTED IN INDEX.HTML.
	// IN THIS CASE, IT FETCHES THE USERS FROM THE SERVER DATABASE, CONVERTS THE RESPONSE TO JSON AND 
	// UPDATES THIS.STATE.ROBOTS WITH THE USERS FETCHED
	componentDidMount() {
		fetch("https://jsonplaceholder.typicode.com/users")
		.then(response => response.json())
		.then(users => this.setState({robots: users}));
	}
	// TRIGGERED EVERY TIME THE SEARCHBOX CHANGES
	onSearchChange = (event) => {
		this.setState({ searchField: event.target.value });
	}
	// RENDER TO INDEX.HTML
	render() {
		const filteredRobots = this.state.robots.filter((robot) => {
		// Needs .toLowerCase() on both sides of the declaration to compare input to user.name on the same level
			return robot.name.toLowerCase().includes(this.state.searchField.toLowerCase()); 
		})
	    	if (!this.state.robots.length) {
	    		return <h1 className="tc f1">Loading...</h1> 
	    	} else {
	      		return (
	      			<div className="tc ">
	      	        	<h1 className="f2">RoboFriends</h1>
	      	        	<SearchBox searchChange={this.onSearchChange}/>
	      	        	<Scroll>
	      	        		<Cardlist robots={filteredRobots}/>
	      	        	</Scroll>
	      	      	</div>
	      	    )
	    	}
	}
}

export default App;
