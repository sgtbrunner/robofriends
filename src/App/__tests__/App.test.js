import { screen, waitForElementToBeRemoved, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { rest } from 'msw';

import App from '../App';
import ROBOTS from '../../mocks/users';
import server from '../../mocks/server';
import { USERS_URL } from '../../utils/api.utils';
import { mountComponent } from '../../utils/test.utils';

const ALERT = 'alert';
const HEADING = 'heading';
const IMAGE = 'img';
const ROBOT = 'robot';
const SEARCHBOX = 'searchbox';

describe('App', () => {
  it('should match snapshot', () => {
    const { container } = mountComponent({ component: App });
    expect(container).toMatchSnapshot();
  });

  it(`loads, renders the robots information and filters cards according
    to user input`, async () => {
    mountComponent({ component: App });

    // shows loading component on start
    const alert = screen.getByRole(ALERT);
    expect(alert).toBeInTheDocument();
    expect(alert).toHaveTextContent(/loading/i);

    // alert no longer appears on page
    await waitForElementToBeRemoved(() => screen.queryByRole(ALERT));

    // // it should render the title
    const title = screen.getByRole(HEADING, { name: /robofriends/i });
    expect(title).toBeInTheDocument();

    // it should render a searchbox
    const searchbox = screen.getByRole(SEARCHBOX);
    expect(searchbox).toBeInTheDocument();

    // it should render the cards
    let cardImages = screen.getAllByRole(IMAGE, { name: ROBOT });
    expect(cardImages).toHaveLength(5);
    ROBOTS.forEach((robot) => {
      const cardName = screen.getByRole(HEADING, { name: robot.name });
      const cardEmail = screen.getByText(robot.email);
      expect(cardName).toBeInTheDocument();
      expect(cardEmail).toBeInTheDocument();
    });

    // it shows two (2) filtered cards when user inputs "super"
    userEvent.clear(searchbox);
    userEvent.type(searchbox, 'super');

    cardImages = screen.getAllByRole(IMAGE, { name: ROBOT });
    expect(cardImages).toHaveLength(2);

    // // it shows one (1) filtered card when user inputs "guard"
    userEvent.clear(searchbox);
    userEvent.type(searchbox, 'guard');

    cardImages = screen.getAllByRole(IMAGE, { name: ROBOT });
    // eslint-disable-next-line jest-dom/prefer-in-document
    expect(cardImages).toHaveLength(1);
  });

  it(`fails to fetch data and there is a problem getting the robots 
  information`, async () => {
    server.resetHandlers(rest.get(USERS_URL, (req, res, ctx) => res(ctx.status(500))));

    mountComponent({ component: App });

    // shows loading alert on start
    const alert = screen.getByRole(ALERT);
    expect(alert).toBeInTheDocument();
    expect(alert).toHaveTextContent(/loading/i);

    // alert changes from loading to error
    await waitFor(() => {
      expect(alert).toHaveTextContent(/an error occurred. please try again later!/i);
    });
  });
});
