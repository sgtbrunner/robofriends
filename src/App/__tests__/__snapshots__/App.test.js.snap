// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<div>
  <div
    class="tc"
  >
    <div
      class="alert-container"
      role="alert"
    >
      <h1
        class="t1 loading"
        style="font-size: 4rem;"
      >
        Loading
      </h1>
    </div>
  </div>
</div>
`;
